apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common.IntegrationTests/BaseEntryPolling.cs

```diff
@@ -55,22 +58,30 @@ protected BaseEntryPolling(TimeSpan timeout = default, TimeSpan sleepInterval =
         /// <returns>The entries or an empty list if not enough entries are found or polling times out.</returns>
         protected IEnumerable<T> GetEntries(int minEntries, Func<IEnumerable<T>> getEntries)
         {
-            TimeSpan totalSleepTime = TimeSpan.Zero;
-            List<T> lastFoundEntries = new List<T>();
-            while (totalSleepTime < _timeout)
+            TimeSpan sleptTime = TimeSpan.Zero;
+            do
             {
-                TimeSpan sleepTime = minEntries > 0 ? _sleepInterval : _timeout;
-                totalSleepTime += sleepTime;
-                Thread.Sleep(sleepTime);
+                var entries = getEntries().ToList();
+                if ((minEntries != 0 && entries.Count >= minEntries) ||
+                    (minEntries == 0 && entries.Count > 0))
+                {
+                    return entries;
+                }
 
-                lastFoundEntries = getEntries().ToList();
-                if (lastFoundEntries.Count >= minEntries)
+                if ((sleptTime += _sleepInterval) < _timeout)
                 {
-                    return lastFoundEntries;
+                    Thread.Sleep(_sleepInterval);
+                }
+                else if (minEntries == 0)
```

Comment:
If minEntries is 0 and there are no entries, we'll just sleep and then return the previously-fetched list anyway. Do we need that sleep? If not, we can simplify the check earlier, and remove this else block.

Reply:
In many cases we use minEntries == 0 to test that no entries were written (because of QPS for instance). Since there's an SLA for entry freshness we really should sleep some to check that we really didn't write any, instead of just not seeing any because they are not available yet.
Last I checked SLA was 10 seconds for 95% of entries and 5 minutes for 99%. We are waiting for 2 minutes in these cases

Reply:
But we're waiting for 2 minutes *and then returning the list from 2 minutes earlier* unless I've missed something.

I do wonder whether for the "no entries" polling we shouldn't just use a different abstraction - it feels like this doesn't really fit.

Reply:
I don't think we are doing that:

In general, how it works for minEntries == 0, we poll at intervals, and as soon as we find at least one entry, we return that. Else, when we reach timeout, instead of throwing, we return an empty list.

Implemented on a do while:
- Fetch the entries
- If we can sleep withing the timeout, sleep -> loop again
- If we couldn't sleep, and we were expecting no entries, we return the list that we just fetch (we couldn't sleep so that list is fresh), which is empty anyways, else, we would have returned it earlier.
- Else we throw as we dind't find as many entries as we were expecting on the timeout.

Or is it me who is missing something?

I did think about separating the minEntries out but it seemed a little bit of a hassle. But I'm happy to reconsider.

Reply:
Ah... I'd ignored that the previous block would be executed for the retry situation. Doh.

Yes, this works - even though it's a little ugly IMO. Fundamentally "poll until a condition is met, with a timeout" and "poll until a timeout, then check a condition" aren't the same thing, and could do with being separated IMO. But that can definitely be done later. Maybe add a TODO at this point?

Reply:
Yep, I agree it's ugly. Added a TODO for consideration when we revisit these.